docs/spec_v6.md

```diff
@@ -0,0 +1,220 @@
+HTTP request body. The encoding is specified by the `Content-Type`
+header.
+
+### Responses
+
+The **Response-Headers** are just normal HTTP response headers. The
+Twirp wire protocol uses the following headers.
+
+* **Content-Type** The value should be either "application/protobuf"
+  or "application/json" to indicate the encoding of the response
+  message. It must match the "Content-Type" header in the request.
+
+The **Request-Body** is the encoded response message contained in the
+HTTP response body. The encoding is specified by the `Content-Type`
+header.
+
+### Example
+
+The following example shows a simple Echo API definition and its
+corresponding wire payloads.
+
+The example assumes the server base URL is "https://example.com/userprefix".
+
+```proto
+syntax = "proto3";
+
+package example.echoer;
+
+service Echo {
+  rpc Hello(HelloRequest) returns (HelloResponse);
+}
+
+message HelloRequest {
+  string message;
+}
+
+message HelloResponse {
+  string message;
+}
+```
+
+**Proto Request**
+
+```
+POST /userprefix/example.echoer.Echo/Hello HTTP/1.1
+Host: example.com
+Content-Type: application/protobuf
+Content-Length: 15
+
+<encoded HelloRequest>
+```
+
+**JSON Request**
+
+```
+POST /userprefix/example.echoer.Echo/Hello HTTP/1.1
+Host: example.com
+Content-Type: application/json
+Content-Length: 27
+
+{"message":"Hello, World!"}
+```
+
+**Proto Response**
+
+```
+HTTP/1.1 200 OK
+Content-Type: application/protobuf
+Content-Length: 15
+
+<encoded HelloResponse>
+```
+
+**JSON Response**
+
+```
+HTTP/1.1 200 OK
+Content-Type: application/json
+Content-Length: 27
+
+{"message":"Hello, World!"}
+```
+
+
+## Errors
+
+Twirp error responses are always JSON-encoded, regardless of
+the request's Content-Type, with a corresponding
+`Content-Type: application/json` header. This ensures that
+the errors are human-readable in any setting.
+
+Twirp errors are a JSON object with the keys:
+
+* **code**: One of the Twirp error codes as a string.
+* **msg**: A human-readable message describing the error
+  as a string.
+* **meta**: (optional) An object with string values holding
+  arbitrary additional metadata describing the error.
+
+Example:
+
+```json
+{
+  "code": "internal",
+  "msg": "Something went wrong"
+}
+```
+
+Example with metadata:
+
+```json
+{
+  "code": "permission_denied",
+  "msg": "Thou shall not pass",
+  "meta": {
+    "target": "Balrog",
+    "power": "999"
+  }
+}
+```
+
+### Error Codes
+
+Twirp errors always include an error code. This code is represented
+as a string and must be one of a fixed set of codes, listed in the
+table below. Each code has an associated HTTP Status Code. When a
+server responds with the given error code, it must set the
+corresponding HTTP Status Code for the response.
+
+| Twirp Error Code    | HTTP Status | Description
+| ------------------- | ----------- | -----------
+| canceled            | 408 | The operation was cancelled.
```

Comment:
The error code is `canceled`, which is the spelling the Go project uses. Consider s/cancelled/canceled/ in the description for consistency :)